# snps_in_genes
A tool for finding what genes SNPs detected with the CFSAN SNP Pipeline reside in and determining if any SNPs are fixed between different groups on a phylogenetic tree

## get_gene_snps.py
1) takes a multi-vcf file generated by the CFSAN SNP Pipeline and the reference in fasta format
2) prokka is required to get an annotation of the reference

## snpnumber4clade.py
Input is specific to the Agona manuscript and the script will need to be manually edited to be applicable to other studies.
Takes an input directory, fasta file (e.g., SNP matrix) and list of ids in a group of interest and finds the fixed SNP differences among the groups if any exist.  
